correct workflow (#86)
OCR-D/ocrd-website: site/en/workflows.md

The following instruction describes all steps of the OCR workflow. Depending on your particular print or rather images not all of those steps will be necessary to obtain good results. Whether a step is required or optional is indicated in the description of each step. 

## Image Optimization (on Page Level)
At first, the image should be prepared for OCR.

> 

Reviewer: Agreed. What concrete steps can we propose here? `ocrd_im6convert` is currently hard to recommend over image magick cmdline or similar if the user knows what steps make sense for the data. But I struggle to give a general recommendation to explore pre-(pre-)processing for possibly better results in certain cases... @VolkerHartmann since you've been benchmarking All The Workflows: Have you experimented with pre-binarization image processing and iff so can you share your experience?

> 

Reviewer 2: > What concrete steps can we propose here?

I would at least document the fact that this can be vital.

> `ocrd_im6convert` is currently hard to recommend over image magick cmdline or similar if the user knows what steps make sense for the data.

What do you mean hard to recommend? It's the only off the shelf solution we can offer. You can see some examples in the ocrd-tool description.

> But I struggle to give a general recommendation to explore pre-(pre-)processing for possibly better results in certain cases

I have experimented a little with the `textcleaner` script from Fred Weinhaus, using

    textcleaner -g -e normalize -f 25 -o 2 -t 30

That is:
1. grayscale conversion 
2. normalize brightness
3. noise filtering
4. text smoothing

which is then equivalent to `-respect-parenthesis -colorspace gray -type grayscale -normalize \( -clone 0 -colorspace gray -negate -lat 25x25x2% -contrast-stretching 0 -blur 1x65535 -level 30x100% \) -compose copy_opacity -composite -fill white -opaque none -alpha off`.

But obviously, the filter param constants I chose may depend on the input's pixel resolution and script height.

>   </tbody>

Author: oh, this was just an auto-completion typo

## Post Correction (Optional)

### Step 15: Text alignment

> In this processing step texts from multiple OCR-engines are aligned in one PAGE.xml.
> 
> **Note:** This step is only required if you want to correct the text afterwards
> by using the output of several OCR-engines as with `ocrd-cis-post-correct.sh`. 
> To obtain good results the input texts should be recognized in the previous step on glyph or at least on word level.

Reviewer 2: ```suggestion
In this processing step, text results from multiple OCR engines (in different annotations sharing the same line segmentation) are aligned into one annotation with `TextEquiv` alternatives.

**Note:** This step is only required if you want to do post-correction afterwards,
feeding alternative character hypotheses from several OCR-engines to improve the search space.
The previous recognition step must be run on glyph or at least on word level.
```

  </tbody>
</table>

> ### Step 16: Post correction

Reviewer 2: ```suggestion
### Step 16: Post-correction
```

> In this processing step you can try to correct the recognized text. 

Reviewer 2: ```suggestion
In this processing step, the recognized text is corrected by statistical error modelling, language modelling, and word modelling (dictionaries, morphology and orthography).
```

> **Note:** `ocrd-cor-asv-ann-process` takes the unaligned OCR text as input, whereas
> `ocrd-cis-post-correct.sh` needs aligned texts.

Reviewer 2: ```suggestion
**Note:** Most tools benefit strongly from input which includes alternative OCR hypotheses. Currently, models for `ocrd-cor-asv-ann-process` are optimised for input from single OCR engines, whereas `ocrd-cis-post-correct.sh` expects input from multi-OCR alignment.
```

  </tbody>
</table>

## Evaluation (Optional)

If Ground Truth data is available, the OCR can be evaluated.

> ### Step 17: Evaluation

Reviewer 2: ```suggestion
### Step 17: OCR Evaluation
```

(as opposed to OLR evaluation, which we don't even have yet)

> In this processing step the output of the OCR can be evaluated.

Reviewer 2: ```suggestion
In this processing step, the text output of the OCR or post-correction can be evaluated by aligning with ground truth text and measuring the error rates.
```

>   </tbody>

Reviewer 2: Wait, where's `ocrd-cor-asv-ann-evaluate`?

Reviewer: It's in there now, even with a tip on how to capture STDERR.

</table>

## Format Conversion (Optional)

OCR-D produces PAGE XML files which contain the recognized text as well as detailed
information on the structure of the processed pages, the coordinates of the recognized
elements etc. Optionally, the PAGE XML can be converted to a different output format.

### Step 18: Format Conversion

In this processing step the produced PAGE XML files can be converted to

> ALTO, hOCR or text files. Note that ALTO and hOCR can also be converted into

Reviewer 2: or PDF! https://github.com/JKamlah/ocrd_pagetopdf is working quite well already, and can be easily _viewed_.

Reviewer: https://github.com/OCR-D/ocrd-website/pull/86/commits/d523926ba3c912179c05cbfa6ee747b11a3c1845

	  <td><code>ocrd-cor-asv-ann-evaluate -I OCR-D-GT,OCR-D-OCR</code></td>
    </tr>
  </tbody>
</table>

## Format Conversion (Optional)

OCR-D produces PAGE XML files which contain the recognized text as well as detailed
information on the structure of the processed pages, the coordinates of the recognized
elements etc. Optionally, the PAGE XML can be converted to a different output format.

### Step 18: Format Conversion

In this processing step the produced PAGE XML files can be converted to ALTO,
PDF, hOCR or text files. Note that ALTO and hOCR can also be converted into
different formats whereas the PDF version of PAGE XML OCR results is a widely
accessible format that can be used as-is by expert and layman alike.

#### Available processors

<table class="processor-table">
<thead>
    <tr>
      <th>Processor</th>
      <th>Parameter</th>
      <th>Remarks</th>
      <th>Call</th>
    </tr>

</thead>
<tbody>
    <tr>
      <td>ocrd-fileformat-transform</td>

>       <td><code>
>         {"from-to": "alto2.0 alto3.0"} 
>         <br># or {from-to: "alto2.0 alto3.1"}
>         <br># or {from-to: "alto2.0 hocr"}
>         <br># or {from-to: "alto2.1 alto3.0"}
>         <br># or {from-to: "alto2.1 alto3.1"}
>         <br># or {from-to: "alto2.1 hocr"}
>         <br># or {from-to: "alto page"}
>         <br># or {from-to: "alto text"}
>         <br># or {from-to: "gcv hocr"}
>         <br># or {from-to: "hocr alto2.0"}
>         <br># or {from-to: "hocr alto2.1"}
>         <br># or {from-to: "hocr text"}
>         <br># or {from-to: "page alto"}
>         <br># or {from-to: "page hocr"}
>         <br># or {from-to: "page text"}
>       </code>

Reviewer 2: ```suggestion
      <td><code><pre>
        {"from-to": "alto2.0 alto3.0"} 
        # or {from-to: "alto2.0 alto3.1"}
        # or {from-to: "alto2.0 hocr"}
        # or {from-to: "alto2.1 alto3.0"}
        # or {from-to: "alto2.1 alto3.1"}
        # or {from-to: "alto2.1 hocr"}
        # or {from-to: "alto page"}
        # or {from-to: "alto text"}
        # or {from-to: "gcv hocr"}
        # or {from-to: "hocr alto2.0"}
        # or {from-to: "hocr alto2.1"}
        # or {from-to: "hocr text"}
        # or {from-to: "page alto"}
        # or {from-to: "page hocr"}
        # or {from-to: "page text"}
      </pre></code>
```

      </td>
      <td>&nbsp;</td>
      <td><code>ocrd-fileformat-transform -I OCR-D-OCR -O OCR-D-ALTO</code></td>
    </tr>

> 

Reviewer 2: ```suggestion
```

This newline might be responsible for the broken table rendering. (I guess the Markdown engine tries to pick it up.)

    <tr>
      <td>ocrd-pagetopdf</td>

>       <td><code>
>       {
>         <br># fix (invalid) negative coordinates
>         <br>"negative2zero": true,
>         <br># create a single "fat" PDF
>         <br>"multipage": true,
>         <br># render text on this level
>         <br>"textequiv_level": "word",
>         <br># outline lines in the PDF
>         <br>"outlines": "line"
>       }
>       </code>

Reviewer 2: ```suggestion
      <td><code><pre>
      {
        # fix (invalid) negative coordinates
        "negative2zero": true,
        # create a single "fat" PDF
        "multipage": true,
        # render text on this level
        "textequiv_level": "word",
        # draw polygon outlines in the PDF
        "outlines": "line"
      }
      </pre></code>
```
